common/java_deps/rules.bzl

```diff
@@ -108,28 +108,30 @@ _transitive_collect_maven_coordinate = aspect(
 
 def _java_deps_impl(ctx):
     names = {}
-    files = []
-    filenames = []
-    outputPathOverrides = ctx.attr.java_deps_root_overrides
+    files_by_output_path = {}
+    output_path_overrides = ctx.attr.java_deps_root_overrides
 
     mapping = ctx.attr.target[TransitiveJarToMavenCoordinatesMapping].mapping
 
-    for file in ctx.attr.target.data_runfiles.files.to_list() + ctx.attr.target.files.to_list():
-        if file.extension == 'jar' and not file.path.startswith(LOCAL_JDK_PREFIX):
+    for file in ctx.attr.target.data_runfiles.files.to_list():
```

Comment:
High level explanation:

For each JAR file, not part of the JDK, that `ctx.attr.target` imports as a runtime dependency:
1. Check that it has a Maven coordinates tag, if this `java_deps` mandates them
2. Determine the output path of this JAR. If it has Maven coordinates, replace `.` and `:` in them to `-`.
3. Add the output path and JAR to a map (of output paths to JARs). If the output filename was already taken by another JAR, throw. The values of this map are the primary outputs of this `java_deps`.
4. Determine the _full_ output path of this JAR, normally `java_deps_root + output_path + ".jar"`, but may be overridden using the `output_path_overrides` map (e.g: to enable some JARs to be included on the classpath only when certain criteria are met). This is used to output a "JAR mapping file" which can be used as a reference.